File: Base/src/main/java/com/flod/android/arch/base/net/log/DefaultHttpLogPrinter.java
package com.flod.android.arch.base.net.log;

import android.text.TextUtils;
import android.util.Log;

import com.flod.android.arch.base.utils.JsonFormatUtil;

import okhttp3.Request;
import okhttp3.Response;

/**
 * SimpleDes:
 * Creator: Flood
 * Date: 2019-03-14
 * UseDes:TODO 两个请求同时到达就会重叠
 */
@SuppressWarnings({"unused", "ConstantConditions"})
public class DefaultHttpLogPrinter implements HttpLogPrinter {
    private static final boolean isFormatBody = true; //是否格式化jsonBody
    private static final String TAG = "HttpLogPrinter";
    private static final String[] LogExtra = {"-\uD83D\uDE03-", "-\uD83D\uDE11-", "-\uD83D\uDE33-", "-\uD83D\uDE06-"};

    private static final String MARK_REQUEST_TOP_LINE = "────── Request ────────────────────────────────────────────────────────────────────────────────────────────";
    private static final String MARK_RESPONSE_TOP_LINE = "────── Response ───────────────────────────────────────────────────────────────────────────────────────────";
    private static final String MARK_CENTER_LINE = "│ ";
    private static final String MARK_BOTTOM_LINE = "───────────────────────────────────────────────────────────────────────────────────────────────────────────";

    private static final String MARK_CORNER_TOP = "┌";
    private static final String MARK_CORNER_CENTER = "├";
    private static final String MARK_CORNER_BOTTOM = "└";

    private static final String MARK_URL = "URL: ";
    private static final String MARK_METHOD = "Method: @";
    private static final String MARK_HEADERS = "Headers: ";
    private static final String MARK_BODY = "Body: ";
    private static final String MARK_STATUS_CODE = "Status Code: ";
    private static final String MARK_TIME = "Time: ";

    private static final String N = "\n";
    private static ThreadLocal<Integer> tagIndex = new ThreadLocal<Integer>() {
        @Override
        protected Integer initialValue() {
            return 0;
        }
    };


    private static void logRequest(Request request, String bodyString) {
        String tag = TAG + "-Request";
        logMulti(tag, MARK_REQUEST_TOP_LINE,
                MARK_URL + request.url(),
                MARK_METHOD + request.method(),
                N,
                MARK_HEADERS,
                formatHeaders(request.headers().toString()),
                N,
                MARK_BODY,
                formatBody(bodyString),
                MARK_BOTTOM_LINE
        );
    }

    private static void logResponse(long chainMs, Response response, String bodyString) {
        String tag = TAG + "-Response";
        logMulti(tag, MARK_RESPONSE_TOP_LINE,
                MARK_URL + response.request().url(),
                MARK_METHOD + response.request().method(),
                MARK_STATUS_CODE + response.code() + " / " + response.message(),
                MARK_TIME + chainMs + "ms", N,
                MARK_HEADERS,
                formatHeaders(response.headers().toString()),
                N,
                MARK_BODY,
                formatBody(bodyString),
                MARK_BOTTOM_LINE);
    }

    private static void log(String tag, String string) {
        Log.d(getTag(tag),string);
    }

    /**
     * 执行多条打印
     *
     * @param tag     传入日志的tag，在tag前会额外添加多余的字符，为了防止日志被合并成一条
     * @param strings 传入的String array
     */
    private static void logMulti(String tag, String... strings) {
        String extra;
        for (int i = 0; i < strings.length; i++) {
            if (i == 0) {
                extra = MARK_CORNER_TOP;
            } else if (i == strings.length - 1) {
                extra = MARK_CORNER_BOTTOM;
            } else {
                extra = MARK_CENTER_LINE;
            }

            String[] childStr = strings[i].split(N);
            if (childStr.length > 0) {
                for (String s : childStr) {
                    log(tag, extra + s);
                }
                continue;
            }
            log(tag, extra + strings[i]);
        }
    }

    private static String formatHeaders(String str) {
        if (TextUtils.isEmpty(str)) {
            return "Empty";
        }

        String[] headers = str.split(N);
        StringBuilder builder = new StringBuilder();
        String extra;
        if (headers.length > 1) {
            for (int i = 0; i < headers.length; i++) {
                if (i == 0) {
                    extra = MARK_CORNER_TOP;
                } else if (i == headers.length - 1) {
                    extra = MARK_CORNER_BOTTOM;
                } else {
                    extra = MARK_CORNER_CENTER;
                }
                builder.append(extra).append(" ").append(headers[i]).append(N);
            }
        } else {
            builder.append("• ").append(str);
        }

        return builder.toString();

    }

    private static String formatBody(String bodyString) {
        if (TextUtils.isEmpty(bodyString))
            return "Omitted Body";
        //bodyString = unicodeToString(bodyString);
        if (isFormatBody)
            return JsonFormatUtil.jsonFormat(bodyString);
        else return bodyString;
    }

/*

    public static String unicodeToString(String str) {
        Pattern pattern = Pattern.compile("(\\\\u(\\p{XDigit}{4}))");
        Matcher matcher = pattern.matcher(str);
        char ch;
        while (matcher.find()) {
            String group = matcher.group(2);
            ch = (char) Integer.parseInt(group, 16);
            String group1 = matcher.group(1);
            str = str.replace(group1, ch + "");
        }
        return str;
    }
*/

    /**
     * 在快速连续打印时日志会把多条合并成一条，导致日志不能对齐，解决方案时是不断改变每条log的tag
     */
    private static String getTag(String tag) {
        int index = tagIndex.get();
        if (index > LogExtra.length - 1) {
            tagIndex.set(0);
            index = 0;
        }
        String extra = LogExtra[index];
        tagIndex.set(++index);
        return extra + tag;
    }

    @Override
    public void printRequest(Request request, String bodyString) {
        logRequest(request, bodyString);
    }

    @Override
    public void printRequestOmitted(Request request) {
        logRequest(request, null);
    }

    @Override
    public void printResponse(long chainMs, Response response, String bodyString) {
        logResponse(chainMs, response, bodyString);
    }

    @Override
    public void printResponseOmitted(long chainMs, Response response) {
        logResponse(chainMs, response, null);
    }
}

File: Base/src/main/java/com/flod/android/arch/base/net/log/DefaultHttpLogPrinter.kt
@file:Suppress("unused", "SpellCheckingInspection")

package com.flod.android.arch.base.net.log

import android.util.Log
import com.google.gson.GsonBuilder
import com.google.gson.JsonParser
import okhttp3.Interceptor
import okhttp3.MediaType
import okhttp3.Request
import okhttp3.Response
import okio.Buffer
import okio.GzipSource
import java.io.IOException
import java.nio.charset.Charset
import java.util.*

/**
 * SimpleDes:
 * Creator: Flood
 * Date: 2019-03-14
 * UseDes:
 */

class DefaultHttpLogPrinter(
    private val logLevel: Level = Level.NONE,
    private val formatBody: Boolean = true
) : Interceptor {

    enum class Level {
        NONE,     //不打印
        REQUEST,  //只打印请求信息
        RESPONSE, //只打印响应信息
        ALL       //打印全部
    }

    companion object {

        private const val TAG = "HttpLogPrinter"
        private val LogTagExtras =
            arrayOf("-\uD83D\uDE03-", "-\uD83D\uDE11-", "-\uD83D\uDE33-", "-\uD83D\uDE06-")
        private const val BOUND_REQUEST_TOP =
            "────── Request ────────────────────────────────────────────────────────────────────────────────────────────"
        private const val BOUND_RESPONSE_TOP =
            "────── Response ───────────────────────────────────────────────────────────────────────────────────────────"
        private const val BOUND_CENTER_LEFT = "│ "
        private const val BOUND_BOTTOM =
            "───────────────────────────────────────────────────────────────────────────────────────────────────────────"
        private const val BOUND_CORNER_TOP = "┌"
        private const val BOUND_CORNER_CENTER = "├"
        private const val BOUND_CORNER_BOTTOM = "└"
        private const val MARK_URL = "URL: "
        private const val MARK_METHOD = "Method: @"
        private const val MARK_HEADERS = "Headers: "
        private const val MARK_BODY = "Body: "
        private const val MARK_STATUS_CODE = "Code: "
        private const val MARK_TIME = "Time: "
        private const val N = "\n"
    }


    private var tagIndex: Int = 0


    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        val logReq = logLevel == Level.ALL || logLevel == Level.REQUEST
        val logRsp = logLevel == Level.ALL || logLevel == Level.RESPONSE

        if (logReq)
            logRequest(request, requestBodyToStr(request))

        val t1 = System.currentTimeMillis()
        val response = chain.proceed(request)
        val t2 = System.currentTimeMillis()
        val chainMs = t2 - t1

        if (logRsp) {
            logResponse(chainMs, response, responseBodyToStr(response))
        }

        return response
    }


    private fun logRequest(request: Request, bodyString: String?) {

        val tag = "$TAG-Request"
        logMulti(
            tag,
            BOUND_REQUEST_TOP,
            MARK_URL + request.url,
            MARK_METHOD + request.method,
            N,
            MARK_HEADERS,
            formatHeaders(request.headers.toString()),
            MARK_BODY,
            formatBody(bodyString),
            BOUND_BOTTOM
        )
    }

    private fun logResponse(chainMs: Long, response: Response, bodyString: String?) {
        val tag = "$TAG-Response"
        logMulti(
            tag,
            BOUND_RESPONSE_TOP,
            MARK_URL + response.request.url,
            MARK_METHOD + response.request.method,
            MARK_STATUS_CODE + response.code + " / " + response.message,
            MARK_TIME + chainMs + "ms",
            N,
            MARK_HEADERS,
            formatHeaders(response.headers.toString()),
            MARK_BODY,
            formatBody(bodyString),
            BOUND_BOTTOM
        )
    }


    private fun log(tag: String, string: String) {
        Log.d(getTag(tag), string)
    }

    /**
     * 执行多条打印
     *
     * @param tag     传入日志的tag，在tag前会额外添加多余的字符，为了防止日志被合并成一条
     * @param strings 传入的String array
     */
    private fun logMulti(tag: String, vararg strings: String) {

        for (i in strings.indices) {
            val extra = when (i) {
                0 -> {
                    BOUND_CORNER_TOP
                }
                strings.size - 1 -> {
                    BOUND_CORNER_BOTTOM
                }
                else -> {
                    BOUND_CENTER_LEFT
                }
            }
            val childStr = strings[i].split(N).toTypedArray()
            if (childStr.isNotEmpty()) {
                for (s in childStr) {
                    log(tag, extra + s)
                }
                continue
            }
            log(tag, extra + strings[i])
        }

    }

    private fun formatHeaders(str: String): String {
        if (str.isBlank()) return "Empty"

        val headers = str.split(N).toMutableList()
        headers.removeAt(headers.size - 1)
        val builder = StringBuilder()
        var extra: String
        if (headers.size > 1) {
            for (i in headers.indices) {
                extra = when (i) {
                    0 -> {
                        BOUND_CORNER_TOP
                    }
                    headers.size - 1 -> {
                        BOUND_CORNER_BOTTOM
                    }
                    else -> {
                        BOUND_CORNER_CENTER
                    }
                }
                builder.append(extra).append(" ").append(headers[i]).append(N)
            }
        } else {
            builder.append("• ").append(str)
        }
        return builder.toString()
    }

    @Suppress("ConstantConditionIf")
    private fun formatBody(bodyString: String?): String {

        if (bodyString.isNullOrBlank()) return "Omitted Body"

        return if (formatBody && bodyString.startsWith("{")) {

            return kotlin.runCatching {
                val gson =
                    GsonBuilder().setPrettyPrinting().disableHtmlEscaping().setLenient().create()
                val jsonElement = JsonParser.parseString(bodyString)
                gson.toJson(jsonElement)
            }.getOrElse {
                log("$TAG-FormatErr", it.stackTraceToString())
                bodyString
            }
        } else bodyString
    }

    /**
     * 在快速连续打印时日志会把多条合并成一条，导致日志不能对齐，解决方案时是不断改变每条log的tag
     */
    private fun getTag(tag: String): String {
        if (tagIndex > LogTagExtras.size - 1) {
            tagIndex = 0
        }
        val extra = LogTagExtras[tagIndex]
        tagIndex++
        return extra + tag
    }


    private fun requestBodyToStr(request: Request): String? {

        if (!isPlaintext(request.body?.contentType())) return null
        val body = request.newBuilder().build().body ?: return null

        return try {
            val buffer = Buffer()
            body.writeTo(buffer)
            //设置读取的字符集类型
            val charset = getCharset(body.contentType())
            buffer.readString(charset)
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }


    /**
     * 从Response获取String
     */
    private fun responseBodyToStr(response: Response): String? {
        val body = response.body ?: return null

        if (!isPlaintext(body.contentType())) return null

        val responseBody = response.newBuilder().build().body ?: return null

        return try {

            //通过response.body直接读取string会直接关闭流,流只能使用一次
            val source = responseBody.source()
            source.request(Long.MAX_VALUE)
            //拿到Buffer
            var buffer = source.buffer
            //获取content的编码类型（可能是gzip）
            val encoding = response.headers["Content-Encoding"]
            if (encoding != null && encoding.equals("gzip", ignoreCase = true)) {
                GzipSource(buffer.clone()).use { gzippedResponseBody ->
                    buffer = Buffer()
                    buffer.writeAll(gzippedResponseBody)
                }
            }
            //设置读取的字符集类型
            val charset = getCharset(responseBody.contentType())
            buffer.clone().readString(charset)
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }


    /**
     * 是否是纯文本格式
     */
    private fun isPlaintext(mediaType: MediaType?): Boolean {
        if (mediaType == null) return false
        if (mediaType.type.toLowerCase(Locale.getDefault()) == "text") {
            return true
        }
        val subtype = mediaType.subtype.toLowerCase(Locale.getDefault())
        return subtype.contains("form")
                || subtype.contains("json")
                || subtype.contains("xml")
                || subtype.contains("html")
                || subtype.contains("x-javascript")
                || subtype.contains("plain")


    }

    private fun getCharset(contentType: MediaType?): Charset {
        return contentType?.charset(Charsets.UTF_8) ?: Charsets.UTF_8
    }
}